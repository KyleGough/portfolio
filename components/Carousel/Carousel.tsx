import React, { useEffect, useRef, useState } from 'react';

import { CarouselImage } from './CarouselImage';

interface CarouselProps {
  children: React.ReactNode;
  backgrounds: string[];
}

export const Carousel: React.FC<CarouselProps> = ({
  children,
  backgrounds,
}) => {
  const [carousel, setCarousel] = useState(0);
  const carouselCount = backgrounds.length;
  const carouselInterval = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    carouselInterval.current = setInterval(carouselNext, 10_000);
    return () => {
      window.clearInterval(carouselInterval.current as NodeJS.Timeout);
    };
  });

  const carouselNext = () =>
    setCarousel((carousel) => (carousel + 1) % carouselCount);

  return (
    <div className="flex justify-center items-center h-screen relative">
      <div className="absolute inset-0 bg-black opacity-70 w-full h-full z-10"></div>
      {backgrounds.map((image, index) => (
        <CarouselImage
          key={`${image}_${index}`}
          bg={image}
          show={carousel === index}
        />
      ))}
      <div className="z-20 absolute inset-0 text-white mx-auto text-center flex tracking-tight items-center">
        {children}
      </div>
    </div>
  );
};
