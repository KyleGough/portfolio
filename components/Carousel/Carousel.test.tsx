import { render, screen } from '@testing-library/react';
import React from 'react';

import { Carousel } from './Carousel';

describe('Carousel component', () => {
  it('renders', async () => {
    const { container } = render(
      <Carousel backgrounds={['bg-1', 'bg-2']}>Carousel Text</Carousel>
    );

    expect(screen.getByText('Carousel Text')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
});
