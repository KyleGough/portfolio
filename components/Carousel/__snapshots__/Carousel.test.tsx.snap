// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component renders 1`] = `
<div>
  <div
    class="flex justify-center items-center h-screen relative"
  >
    <div
      class="absolute inset-0 bg-black opacity-70 w-full h-full z-10"
    />
    <div
      class="opacity-100 absolute top-0 left-0 w-full h-full clip-inset duration-2000 transition-opacity"
    >
      <img
        alt=""
        class="fixed object-cover object-center w-full h-full"
        data-nimg="1"
        decoding="async"
        height="100"
        src="/_next/image?url=%2F%23image&w=256&q=75"
        srcset="/_next/image?url=%2F%23image&w=128&q=75 1x, /_next/image?url=%2F%23image&w=256&q=75 2x"
        style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
        width="100"
      />
    </div>
    <div
      class="opacity-0 absolute top-0 left-0 w-full h-full clip-inset duration-2000 transition-opacity"
    >
      <img
        alt=""
        class="fixed object-cover object-center w-full h-full"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2F%23image&w=256&q=75"
        srcset="/_next/image?url=%2F%23image&w=128&q=75 1x, /_next/image?url=%2F%23image&w=256&q=75 2x"
        style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
        width="100"
      />
    </div>
    <div
      class="z-20 absolute inset-0 text-white mx-auto text-center flex tracking-tight items-center"
    >
      Carousel Text
    </div>
  </div>
</div>
`;
